rust/src/container.rs

```diff
@@ -14,3 +35,359 @@ pub async fn entrypoint(args: &[&str]) -> Result<i32> {
+    dir: &gio::File,
+    ts: &crate::ffi::RpmTs,
+    state: &mut MappingBuilder,
+) -> Result<()> {
+    use std::collections::btree_map::Entry;
+    let cancellable = gio::NONE_CANCELLABLE;
+    let e = dir.enumerate_children(
+        "standard::name,standard::type",
+        gio::FileQueryInfoFlags::NOFOLLOW_SYMLINKS,
+        cancellable,
+    )?;
+    for child in e {
+        let childi = child?;
+        let name: Utf8PathBuf = childi.name().try_into()?;
+        let child = dir.child(&name);
+        path.push(&name);
+        match childi.file_type() {
+            gio::FileType::Regular | gio::FileType::SymbolicLink => {
+                let child = child.downcast::<ostree::RepoFile>().unwrap();
+
+                // Remove the skipped path, since we can't hit it again.
+                if state.skip.remove(Utf8Path::new(path)) {
+                    path.pop();
+                    continue;
+                }
+
+                let mut pkgs = ts.packages_providing_file(path.as_str())?;
+                // Let's be deterministic (but _unstable because we don't care about behavior of equal strings)
+                pkgs.sort_unstable();
+                // For now, we pick the alphabetically first package providing a file
+                let mut pkgs = pkgs.into_iter();
+                let pkgid = pkgs
+                    .next()
+                    .map(|v| -> Result<_> {
+                        // Safety: we should have the package in metadata
+                        let meta = state.packagemeta.get(v.as_str()).ok_or_else(|| {
+                            anyhow::anyhow!("Internal error: missing pkgmeta for {}", &v)
+                        })?;
+                        Ok(Rc::clone(&meta.identifier))
+                    })
+                    .transpose()?
+                    .unwrap_or_else(|| Rc::clone(&state.unpackaged_id));
+                // Track cases of duplicate owners
+                match pkgs.len() {
+                    0 => {}
+                    _ => {
+                        state.multi_provider.push(path.clone());
+                    }
+                }
+
+                let checksum = child.checksum().unwrap().to_string();
+                match state.content.entry(checksum) {
+                    Entry::Vacant(v) => {
+                        v.insert(pkgid);
+                    }
+                    Entry::Occupied(_) => {
+                        let checksum = child.checksum().unwrap().to_string();
+                        let v = state.duplicates.entry(checksum).or_default();
+                        v.push(pkgid);
+                    }
+                }
+            }
+            gio::FileType::Directory => {
+                build_mapping_recurse(path, &child, ts, state)?;
+            }
+            o => anyhow::bail!("Unhandled file type: {}", o),
+        }
+        path.pop();
+    }
+    Ok(())
+}
+
+fn gv_nevra_to_string(pkg: &glib::Variant) -> String {
```

Comment:
Pretty sure we have this already on the C side. Might be worth making this public and moving over those callers.